pkg/libp2p/utils.go

```diff
@@ -75,3 +81,77 @@ func GeneratePrivateKey(numBits int) (crypto.PrivKey, error) {
 	}
 	return prvKey, nil
 }
+
+// upgradeAddress accepts a non p2p multiaddress that denotes a http or https connection
+// and attempts to use it to call the requester node to ask for it's peer id.  If
+// successful it will then convert the http(s) multiaddress into a valid p2p multiaddress
+// and return it for use by the caller.
+func upgradeAddress(ctx context.Context, address multiaddr.Multiaddr) (multiaddr.Multiaddr, error) {
+	type PeerInfo struct {
+		ID string `json:"ID"`
+	}
+
+	type PeerInfoResponse struct {
+		Peerinfo PeerInfo `json:"PeerInfo"`
+	}
+
+	parts := strings.Split(address.String()[1:], "/")
+	ipAddr, port, scheme := parts[1], parts[3], parts[4]
+	url := fmt.Sprintf("%s://%s:%s/api/v1/agent/node", scheme, ipAddr, port)
+
+	client := http.Client{Timeout: time.Second * 2}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return address, err
+	}
+
+	res, err := client.Do(req)
+	if err != nil {
+		return address, err
+	}
+
+	defer func() {
+		if res.Body != nil {
+			res.Body.Close()
+		}
+	}()
+
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return address, err
+	}
+
+	response := PeerInfoResponse{}
+	if err = json.Unmarshal(body, &response); err != nil {
+		return address, nil
+	}
+
+	if response.Peerinfo.ID != "" {
+		// Strip the http(s) from the end of the address
+		if addr, err := multiaddr.NewMultiaddr(fmt.Sprintf("/%s", scheme)); err != nil {
+			return address, err
+		} else {
+			address = address.Decapsulate(addr)
+		}
+
+		// Swap the port to a hard-coded value until we have a better mechanism
+		// for specifying the P2P listen port
+		parts = strings.Split(address.String(), "/")
```

Comment:
Multiaddr has Split and Join functions that might be more helpful and safer than dealing with their string values

Reply:
Fixed in 9deb50ec